### DISCUSSION TITLE
Programmatic access to information

### DISCUSSION THREAD

**Discussion by opk12** [8go/matrix-commander#98] [Ideas]
What about a central place of discussion about the changes that involve at least these (what else?)
1. getting the homeserver's verbatim output (follows the Client-Server API spec)
2. getting custom output (only the fields of interest to the client)
3. getting machine-readable output
4. calling from Python code

So this issue covers at least these (what else?)
- #92
- #95
- ...

**Comment by opk12** [comment]
What about exporting some specific functions from matrix_commander and moving CLI handling into `cli.py`. So one can do this

```
rooms: Iterable[str] = matrix_commander.get_joined_rooms(room_id=...))
json.dumps(rooms)
```

**Comment by 8go** [comment]
Thanks @opk12 for raising the awareness for these issues.

First question: Bullet 1 vs Bullet 3: Are there cases where the homeserver's verbatim output is not machine-readable output?

In other words, if bullet 1 is taken care of, then bullet 3 is automatically resolved as well?

**Reply by opk12** [reply]
No / Yes, even in future spec versions, otherwise one could not make a client software.

**Comment by opk12** [comment]
The pro of the matrix-nio API is that it mirrors the Client-Server spec'ed workflow. The con is that each user must learn the C-S spec and reimplement the same patterns.

De facto, matrix-commander is a public repository of shared patterns / recipes around matrix-nio. It is even promoted as usage example in a dedicated section in the matrix-nio docs ([link](https://matrix-nio.readthedocs.io/en/latest/examples.html#further-reading-and-exploration)). But the CLI models the most common end-user use cases.

I think about a mixed approach: when the spec advances (leaving m-c behind), the use-case is custom, the CLI does not exist, ... then I implement the parts of the spec of my interest, and leave everything else to m-c (reusing m-c as a simple wrapper around matrix-nio). I can safely intermix calling the m-c CLI, calling the m-c functions and classes, parsing the homeserver's output downloaded by m-c.
 
That's why I used very abstract words `output`, `information`, even though the issues open today mention JSON room events.

The feature to send custom JSON events (already implemented) fits as part of this view.

**Reply by opk12** [reply]
Notably, it would be cool if m-c exposed the `AsyncClient` so that I can manipulate it directly.
